File: extra-code/stage2_code/py_srv/py_srv/add_two_ints_server.py
# ============================================================
# Service 通信示例：服务端（Server）
# 演示概念：create_service、request/response 请求-响应模型
# 提供 /add_two_ints 服务：收到两个整数 a、b，返回它们的和
# 运行方式：ros2 run py_srv add_two_ints_server
# ============================================================
import rclpy
# AddTwoInts 接口定义：Request 含 a、b 两个整数，
# Response 含 sum 一个整数
from example_interfaces.srv import AddTwoInts
from rclpy.executors import ExternalShutdownException
from rclpy.node import Node


class AddTwoIntsServer(Node):
    """Provide the /add_two_ints service."""

    def __init__(self):
        super().__init__("add_two_ints_server")
        # 创建 Service：参数是接口类型、服务名、处理回调；
        # 之后客户端每发一次请求就触发一次回调
        self.service_ = self.create_service(
            AddTwoInts, "add_two_ints", self.callback_add_two_ints)
        self.get_logger().info("Add Two Ints server has been started.")

    def callback_add_two_ints(self, request, response):
        # Service 回调固定签名：收到 request，填好 response
        # 并 return，框架会把它发回给客户端
        response.sum = request.a + request.b
        self.get_logger().info(f"{request.a} + {request.b} = {response.sum}")
        return response


def main(args=None):
    rclpy.init(args=args)
    node = AddTwoIntsServer()
    # 服务端必须一直 spin 才能持续响应客户端请求
    try:
        rclpy.spin(node)
    except (KeyboardInterrupt, ExternalShutdownException):
        pass
    finally:
        node.destroy_node()
        if rclpy.ok():
            rclpy.shutdown()


if __name__ == "__main__":
    main()

File: extra-code/stage2_code/cpp_srv/src/add_two_ints_server.cpp
// ============================================================
// Service 通信示例:服务端(Server)
// 演示:提供 /add_two_ints 服务,收到两个整数 a、b 后返回它们的和
// Service 是"一问一答"模式:Client 发 Request,Server 回 Response(区别于 Topic 的持续广播)
// 运行:ros2 run cpp_srv add_two_ints_server
// 测试:ros2 service call /add_two_ints example_interfaces/srv/AddTwoInts "{a: 2, b: 3}"
// ============================================================

#include <memory>

// Service 接口头文件:AddTwoInts 定义了 Request(a, b)和 Response(sum)两部分
#include "example_interfaces/srv/add_two_ints.hpp"
#include "rclcpp/rclcpp.hpp"

// 类型别名,后面写 AddTwoInts 就不用重复冗长的完整命名空间
using AddTwoInts = example_interfaces::srv::AddTwoInts;

class AddTwoIntsServer : public rclcpp::Node
{
public:
  AddTwoIntsServer() : Node("add_two_ints_server")
  {
    // create_service:服务名 + 回调。回调有两个参数(request 和 response),
    // 所以 std::bind 需要两个占位符 _1、_2 分别对应它们
    service_ = create_service<AddTwoInts>(
      "add_two_ints",
      std::bind(&AddTwoIntsServer::callback, this, std::placeholders::_1, std::placeholders::_2));
    RCLCPP_INFO(get_logger(), "Add Two Ints server has been started.");
  }

private:
  // Service 回调:框架传入已填好的 request,我们把结果写进 response 即可,
  // 无需手动"发送"——回调返回后框架自动把 response 发回 Client
  void callback(const AddTwoInts::Request::SharedPtr request,
                const AddTwoInts::Response::SharedPtr response)
  {
    response->sum = request->a + request->b;
    RCLCPP_INFO(get_logger(), "%ld + %ld = %ld", request->a, request->b, response->sum);
  }

  rclcpp::Service<AddTwoInts>::SharedPtr service_;
};

int main(int argc, char ** argv)
{
  rclcpp::init(argc, argv);
  // Server 靠 spin 事件循环等待请求:请求到达时才触发回调
  rclcpp::spin(std::make_shared<AddTwoIntsServer>());
  rclcpp::shutdown();
  return 0;
}

